plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/compare/ComparePresenter.java

```diff
@@ -75,45 +80,125 @@ public ComparePresenter(AppContext appContext,
     }
 
     /**
-     * Show compare window.
+     * Show compare window for given set of files between given revision and HEAD.
      *
-     * @param file
-     *         file name with its full path
-     * @param status
-     *         status of the file
+     * @param alteredFiles
+     *         ordered touched files
+     * @param currentFile
+     *         file which will be shown first, if null then the first from the list will be shown
      * @param revision
      *         hash of revision or branch
      */
-    public void showCompareWithLatest(final File file, final Status status, final String revision) {
-        this.comparedFile = file;
+    public void showCompareWithLatest(final AlteredFiles alteredFiles,
+                                      @Nullable final String currentFile,
+                                      final String revision) {
+        this.alteredFiles = alteredFiles;
         this.revision = revision;
+
         this.compareWithLatest = true;
 
-        if (status.equals(ADDED)) {
-            showCompare("");
-            return;
-        }
+        findCurrentFile(currentFile);
+        showCompareForCurrentFile();
+    }
+
+    /**
+     * Shows compare window for given set of files between specified revisions.
+     *
+     * @param alteredFiles
+     *         ordered touched files
+     * @param currentFile
+     *         file which will be shown first, if null then the first from the list will be shown
+     * @param revisionA
+     *         hash of the first revision or branch.
+     *         If it is set to {@code null}, compare with empty repository state will be performed
+     * @param revisionB
+     *         hash of the second revision or branch.
+     *         If it is set to {@code null}, compare with latest repository state will be performed
+     */
+    public void showCompareBetweenRevisions(final AlteredFiles alteredFiles,
+                                            @Nullable final String currentFile,
+                                            @Nullable final String revisionA,
+                                            @Nullable final String revisionB) {
+        this.alteredFiles = alteredFiles;
+        this.revisionA = revisionA;
+        this.revisionB = revisionB;
+
+        this.compareWithLatest = false;
 
-        final Container rootProject = GitUtil.getRootProject(file);
+        findCurrentFile(currentFile);
+        showCompareForCurrentFile();
+    }
+
+    /**
+     * Shows comparison for selected file.
+     * Type of comparison to show depends on {@code compareWithLatest} field.
+     */
+    private void showCompareForCurrentFile() {
```

Comment:
I thought about it. On the one hand it would be good to generalize `gitDir` and `relPath` and invoke `showCompare*` form a single place, but on the other hand I tired to leave some 'frame' for subrepository support which should be dropped by generalization.